cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewManager.java

```diff
@@ -16,55 +16,259 @@
+  private final ArtifactRepository artifactRepository;
+  private final ArtifactStore artifactStore;
+  private final AuthorizerInstantiator authorizerInstantiator;
+  private final StreamAdmin streamAdmin;
+  private final StreamCoordinatorClient streamCoordinatorClient;
+  private final PrivilegesManager privilegesManager;
+  private final AuthorizationEnforcer authorizationEnforcer;
+  private final Cache<ApplicationId, Injector> appInjectors;
 
-  @Override
-  public void stop(ApplicationId preview) throws Exception {
+  @Inject
+  DefaultPreviewManager(final CConfiguration cConf, Configuration hConf, DiscoveryService discoveryService,
+                        @Named(DataSetsModules.BASE_DATASET_FRAMEWORK) DatasetFramework datasetFramework,
+                        PreferencesStore preferencesStore, SecureStore secureStore,
+                        TransactionManager transactionManager, ArtifactRepository artifactRepository,
+                        ArtifactStore artifactStore, AuthorizerInstantiator authorizerInstantiator,
+                        StreamAdmin streamAdmin, StreamCoordinatorClient streamCoordinatorClient,
+                        PrivilegesManager privilegesManager, AuthorizationEnforcer authorizationEnforcer) {
+    this.cConf = cConf;
+    this.hConf = hConf;
+    this.datasetFramework = datasetFramework;
+    this.discoveryService = discoveryService;
+    this.preferencesStore = preferencesStore;
+    this.secureStore = secureStore;
+    this.transactionManager = transactionManager;
+    this.artifactRepository = artifactRepository;
+    this.artifactStore = artifactStore;
+    this.authorizerInstantiator = authorizerInstantiator;
+    this.streamAdmin = streamAdmin;
+    this.streamCoordinatorClient = streamCoordinatorClient;
+    this.privilegesManager = privilegesManager;
+    this.authorizationEnforcer = authorizationEnforcer;
 
+    // TODO make maximum size and expire after write configurable?
+    this.appInjectors = CacheBuilder.newBuilder()
+      .maximumSize(10)
+      .expireAfterWrite(15, TimeUnit.MINUTES)
+      .removalListener(new RemovalListener<ApplicationId, Injector>() {
+        @Override
+        @ParametersAreNonnullByDefault
+        public void onRemoval(RemovalNotification<ApplicationId, Injector> notification) {
+          Injector injector = notification.getValue();
+          if (injector != null) {
+            PreviewRuntimeService service = injector.getInstance(PreviewRuntimeService.class);
+            service.stopAndWait();
+          }
+          ApplicationId application = notification.getKey();
+          if (application == null) {
+            return;
+          }
+          java.nio.file.Path previewDirPath = Paths.get(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                                                        "preview", application.getApplication()).toAbsolutePath();
+
+          try {
+            DirUtils.deleteDirectoryContents(previewDirPath.toFile(), false);
```

Comment:
If have to convert the `Path` to `File`, why not just construct the `File` directly? Also, seems like you can have the preview directory as a field.